[WIP] Added implementation of SGHMC and example.

dwadden opened the pull request:
I'm working with @nfoti and @yianma. We have some questions on a few small parts of the code. See comments in changed files.

dwadden commented on edward/inferences/sghmc.py:
```diff
@@ -0,0 +1,152 @@
+    decreases.
+    Implements the update equations from (15) of Chen et al., 2014.
+    """
+    # TODO: Would it be more memory-efficient to use tf.control_dependencies,
```
Does making aliases make copies of the tensors? If so, would it save memory to use `tf.control_dependencies` and mutate the objects directly?

dustinvtran replied:
i'm not sure i understood the question but for example:
```python
>>> x = tf.constant(0)
>>> y = x
>>> x, y
(<tf.Tensor 'Const:0' shape=() dtype=int32>, <tf.Tensor 'Const:0' shape=() dtype=int32>)
```
y has a reference to the same tensor and doesn't recreate it from scratch, if that's what you're asking.

dwadden replied:
Yep, that was the question. Just wanted to make sure it wasn't recreating.

dwadden commented on edward/inferences/sghmc.py:
```diff
@@ -0,0 +1,152 @@
+
+    # Simulate Hamiltonian dynamics with friction.
+    friction = tf.constant(self.friction, dtype = tf.float32)
+    # TODO: Allow option for exponentially decaying learning rate, or similar.
```
For SGLD, we found that using a constant learning rate performed better than using exponential decay. Perhaps this should be an option passed into the inference method?

dustinvtran replied:
sure, that makes sense. note one difficulty i had when implementing SGLD was that i could not leverage any of the existing tensorflow optimizers, and if more learning rates were an option, one would need to implement them all from scratch. if there were a way to use the tensorflow optimizers, it would be nice so long as the implementation doesn't ruin the readability and/or speed of the code.

dwadden replied:
Got it re: tensorflow.

I'll make a new PR that adds an option to keep the learning rate constant, along with an example showing cases where this may be a bit easier to work with.

dwadden commented on edward/inferences/sghmc.py:
```diff
@@ -0,0 +1,152 @@
+    z_sample : dict
+      Latent variable keys to samples.
+    """
+    # TODO: This appears to be identical across HMC, SGHMC, SGLD. Should it be
```
The `_log_joint` function looks like it's the same. Refactor?

dustinvtran replied:
yeah, that sounds useful to refactor (although i'm not sure where the best location is to put it)

dwadden replied:
OK. Will think about this and submit another PR.

dwadden commented on examples/normal_sghmc.py:
```diff
@@ -0,0 +1,80 @@
+from edward.models import Empirical, MultivariateNormalFull
+
+# Helper functions
+# TODO: Where should these go?
```
I added some helper functions to visualize the contours of a bivariate Gaussian (or more generally, any Edward distribution with a pdf method). Should this be refactored?

dustinvtran replied:
that's awesome. maybe we can leave it here for now and think about refactoring it in another PR that focuses more on general visualizations for edward

dwadden replied:
Cool.

dwadden commented on examples/normal_sghmc.py:
```diff
@@ -0,0 +1,80 @@
+    if label:
+        plt.clabel(cs, inline=1, fontsize=10)
+
+def retrieve_samples(qz):
```
It took a bit of work to figure out how to grab the actual MCMC samples. Is there a method to do this that we missed, or should we move this elsewhere to allow for more general use?

dustinvtran replied:
if you just want the samples you can do `qz.params`

dwadden replied:
Nice, thanks

dwadden commented:
Did a bit of code cleanup - fixed indentation to 2 spaces in `normal_sghmc.py` and got rid of the TODO's since we've now addressed how to handle them.
Made a minor mistake and squashed those commits into the previous one so that the original changes we discussed don't show up. Now I know not to do that.
Let me know if there's anything else you'd like changed. Otherwise I'll start on the PR for adaptive step sizes.

dwadden commented:
Cleaned things up a bit. Got rid of the README notes since we've now discussed them. Squashed into previous commit, which I now know is a bad thing to do because the changes we discussed are gone. Lesson learned.

Let me know if there are other things you'd like changed on this PR. Otherwise, I'll start thinking about the other thing we discussed: step size, code viz, and refactor of _log_joint.

dustinvtran commented:
looks good to me. can you add an example using data? e.g., linear/logistic regression is fine.

dwadden commented:
Will do.

dwadden commented:
Git style question: I'm want to update this branch with the latest changes from master, which contains relevant bug fixes for #412 . I'm also going to add my integer type fix. 

In general, this can be done by merging master into this branch, or rebasing the branch onto master. The problem with the former approach is it will make the history messy, while doing the latter makes it impossible to view changes that were made before the rebase (like what's happened above).

Which do you prefer?

dustinvtran commented:
i personally prefer to rebase to keep the commit history clean. that said, we're always squashing+merging pull requests, so feel free to do either.

dwadden commented:
Cool, rebase works for me.

dustinvtran commented:
i ran the examples in tensorflow 0.11.0 with python 2 and 3. i also double-checked upon merging that they work in tensorflow 1.0.0rc0. merging now.
